libc-bottom-half/cloudlibc/src/libc/unistd/write.c

```diff
@@ -2,11 +2,131 @@
 //
 // SPDX-License-Identifier: BSD-2-Clause
 
+#ifdef __wasilibc_use_wasip2
+#include <wasi/wasip2.h>
+#include <wasi/descriptor_table.h>
+#include <common/errors.h>
+#include <time.h>
+#else
 #include <wasi/api.h>
+#endif
 #include <errno.h>
 #include <unistd.h>
 
 ssize_t write(int fildes, const void *buf, size_t nbyte) {
+#ifdef __wasilibc_use_wasip2
+  streams_borrow_output_stream_t output_stream;
+  filesystem_own_input_stream_t input_stream;
+  bool ok = false;
+  bool create_new_stream = false;
+  filesystem_error_code_t error_code;
+  descriptor_table_entry_t* entry = 0;
+
+  // Check for stdin/stdout/stderr
+  if (fildes == 1)
+    output_stream = streams_borrow_output_stream(stdout_get_stdout());
+  else if (fildes == 2)
+    output_stream = streams_borrow_output_stream(stderr_get_stderr());
```

Comment:
Can we cache these streams, as well, so that libc only needs to call get_stdout and get_stderr once for the lifetime of the instance?

Reply:
We don't have to worry about dropping the stdio streams or their pollables, those do need to live the life of the instance. But, we should initialize them lazily and just once for the life of the instance.

Reply:
Ah, got it. Done in [a348119](https://github.com/WebAssembly/wasi-libc/pull/606/commits/a348119be29c7cff0d5ad21e9b1d00025875725c)